### PR TITLE
Support for the structured products in the registry

### LINKED ISSUE #632: Add Initial Support for Searching Full PDS4 Structured Metadata
### Are you sure this is not a new requirement or bug?

Yes

### Task Type

Theme

### 💡 Description

For many use cases, searching a flattened version of the data loses some of the ability to search for attributes that are many levels deep. We need the ability to search for these specific attributes.

Additional sub-tasks that need tickets:

- [x] ~Update `ops` data model to be a full LDD~ to be done in separate task
- [x] Update harvest to support new `ops` data model
- [x] Update harvest to support new design
- [x] Update registry-mgr to support new design
- [x] ~Develop sweeper to update/reload all data into newly designed registry~ separate them identified for this.
- [ ] Update registry-api to support searching in this manner

https://docs.google.com/presentation/d/1JLSioueXS38YLGf5IG5BaD52Ytf-ilxp/edit?usp=sharing&ouid=111291853515250351941&rtpof=true&sd=true

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
In order for the registry-api to use the structured field names, need to quit doing the property name conversions. Made it selectable via the properties file just because the transition from one to the other may take a while. This gives us the freedom to use the same API code for older or newer registry-apis.

## ⚙️ Test Data and/or Report
All postman tests pass.

## ♻️ Related Issues
Last step. Closes #632 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Documentation
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.

### Maintenance
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue or Jira Ticket.
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

File: service/src/main/resources/application.properties [hunk 1]
```diff
@@ -51,3 +51,7 @@ filter.archiveStatus=archived,certified
 # source version from maven
 # need to be updated with actual value when runs outside of maven
 registry.service.version=@project.version@
+
+# signal if the database being used is "flat" or "structured"
+# used an enum of strings, "flat" or "structured" for future new types
+registry.field.name.architecture=structured
```
Comment (al-niessner) [inline service/src/main/resources/application.properties:57]: @tloubrieu-jpl 

This is how you want to change between flat and structured DB. I am not sure if the postman tests will pass unless you have jumped them to structured already.

File: service/src/main/java/gov/nasa/pds/api/registry/model/SearchUtil.java [hunk 2]
```diff
@@ -8,17 +8,27 @@
...
 
+  private static String fnArch;
+  @Value("${registry.field.name.architecture}")
+  public void setFnArch(String fnArch) {
+    SearchUtil.fnArch = fnArch;
+  }
+
   static public String jsonPropertyToOpenProperty(String jsonProperty) {
-    return jsonProperty.replace(".", "/");
+    if (SearchUtil.fnArch == null || SearchUtil.fnArch.equalsIgnoreCase("flat")) return jsonProperty.replace(".", "/");
```
Comment (al-niessner) [inline service/src/main/java/gov/nasa/pds/api/registry/model/SearchUtil.java:30]: @jordanpadams @tloubrieu-jpl 

I made the default (older registry-api applications.properties file) behave as flat. We will need to change the == to != and the || to && if you want the missing property condition to be structured instead.

**Comment by al-niessner** [comment]
@nutjob4life @tloubrieu-jpl 

Why is sonar stuff now hidden? Not very useful that it tells me there is one new problem then will not tell me what it is. Should I just start randomly guessing?

**Comment by nutjob4life** [comment]
@al-niessner I'm not sure what's being hidden? I can visit the URL https://sonarcloud.io/dashboard?id=NASA-PDS_registry-api&pullRequest=728 — and I don't have to log in

**Comment by al-niessner** [comment]
My resolution of URL says 1 new problem. A box with red X that says failed.
No like to mor details. Nothing. I can send a screenshot.

**Comment by nutjob4life** [comment]
[So strange…no idea why it doesn't load](https://sonarcloud.io/summary/new_code?id=NASA-PDS_registry-api&pullRequest=728) 🤷

**Comment by nutjob4life** [comment]
[No idea why it doesn't load](https://sonarcloud.io/summary/new_code?id=NASA-PDS_registry-api&pullRequest=728) 🤷

**Comment by tloubrieu-jpl** [comment]
Hi @al-niessner , we don't want to merge that in `develop` I believe because we have other updates which need to go to develop for an earlier release. I creted a 'structured' branch which will be the target branch for this PR.

**Comment by al-niessner** [comment]
@nutjob4life 

Different browser. I can see it in this browser. Anyway, gonna have to live with it as it is an artifact of springboot.

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

It is written to work in both by changing the java property. If the develop is the flat branch then just change java property to flat. When merged into structured branch, change it to structured.

Maybe I do not understand your development idea. Are develop and structured always going to remain apart or will structured be merged into develop? Anyway, I will be in the meeting tomorrow so we can cover it then.

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

I do not understand the build failure that registry-loader-test-init did not complete successfully. Did you want me to dig into the cause?

I see you moved the merge into branch to structured already. Thanks.

**Review by tloubrieu-jpl — APPROVED** [review]
Thanks @al-niessner , I changed the target branch to main as the behavior would stay the same with the current applications.properties.
We'll test in integration with the structured behavior with the same code but updated application.properties after we are done with the current build release (flat behavior).

**Review by tloubrieu-jpl — CHANGES_REQUESTED** [review]
Actually, I see the sonarcloud comment which should be easy to fix:

Image

I had to log in using my github account to see the error.

Once you re-commit the integration test would re-run, we will see if we are more lucky this time. The tests have been running for 6 hours on the other PR and are not done yet. That is weird too.

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

I know what those are and they have nothing to do with this PR. There should be a very large number of functions like that and sonarqube is just off its rocker. They are, as the class says, in the Mock that you created. They have to be implemented to make the interface happy, but we are not testing them today. They are good place holders for when we find out they need tested so doing either suggestion is a tad pointless in a test Mock.

Also, this PR does not touch that file. Not sure why sonarqube is crying about it now.

**Comment by tloubrieu-jpl** [comment]
Thanks @al-niessner . Sonarcube might have changed their rules ?
I could do that, but is there a way to silence these warnings, as they might hide some hypothetical/future relevant one next time ?

**Comment by al-niessner** [comment]
@tloubrieu-jpl 

It has not changed its rules. Mr Gemini says that you want the sonarqube admin to ignore the mock files:

https://share.google/aimode/nlpCtoTD08Em8qM5l

**Comment by tloubrieu-jpl** [comment]
Thanks @al-niessner,  I added the comments recommended by SonarCube. They were missing on 2 new methods.
Let's cross our fingers for the integration tests.

**Comment by jordanpadams** [comment]
@al-niessner can you take a look at the PR comments from Copilot?

File: terraform/ecs.tf [hunk 3]
```diff
@@ -86,11 +86,7 @@ resource "aws_ecs_cluster" "pds-registry-api-ecs" {
   }
 }
 
-# Do we need individual dev/test/prod repositories?
-# I don't think we do, but then we need to use prod account instead of the dev account, would that work ?
-data "aws_ecr_repository" "pds-registry-api-service" {
-  name = "pds-registry-api-service"
-}
+# ECR repository is now defined in ecr.tf
```
Comment (tloubrieu-jpl) [inline terraform/ecs.tf]: Hi @al-niessner , I don't know why we have those changes in the terraform script in this PR ? That would be cleaner to remove them.

**Review by tloubrieu-jpl — CHANGES_REQUESTED** [review]
Hi @al-niessner ,

I am not sure if I made the review before, but now I am seeing terraform updates which don't need to be in this PR. Could you remove them ?

I would like to finally test the version of the code with the flat option to make the release of the registry-api for the build 17.

Thanks

**Review by tloubrieu-jpl — APPROVED** [review]
Test passed in the flat version. Structured version not tested yet.
Code review ok
